### PR TITLE
Support for non-ASCII characters in code (within block structures) for #226

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
Good news: with this branch of fparser, the master of PSyclone copes with all of the NEMO source files in the configuration (with the exception of those that we deliberately ignore).
Bad news: the tofortran() method of Component_Part isn't covered by the test suite and I've modified it.

**Comment by arporter** [comment]
This is ready for first review. I've gone for a minimal change rather than fully implement all of the tests for the various Rules that have been touched (since that touch is tiny).

**Comment by arporter** [comment]
Just a reminder @rupertford that this is pending if you find yourself at a loose end :-)

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -1784,9 +1795,13 @@ def match(string):
         i = line.find('::')
         if i == -1:
             return
-        return p, Proc_Component_Attr_Spec_List(
-            repmap(line[:i].rstrip())), Proc_Decl_List(
-                repmap(line[i+2:].lstrip()))
+        attr_spec_list = Proc_Component_Attr_Spec_List(
+            repmap(line[:i].rstrip()))
+        # C449 POINTER must be present in the attribute list
```
Comment (rupertford) [inline src/fparser/two/Fortran2003.py]: Please sumarise all constraints associated with R445 in the docstring and point out that the other constraints are currently not tested (see for example R451)

File: src/fparser/two/Fortran2003.py [hunk 2]
```diff
@@ -4970,6 +4995,7 @@ def tofortran(self, tab='', isfix=None):
             else:
                 tmp.append(item.tofortran(tab=tab+'  ', isfix=isfix))
         tmp.append(end.tofortran(tab=tab, isfix=isfix))
+        py2_encode_list_items(tmp)
```
Comment (rupertford) [inline src/fparser/two/Fortran2003.py]: Add a test for this as was done for r438 and r808?

File: src/fparser/two/Fortran2003.py [hunk 3]
```diff
@@ -5396,6 +5432,7 @@ def tofortran(self, tab='', isfix=None):
...
```
Comment (rupertford) [inline src/fparser/two/Fortran2003.py]: Add a test for this as was done for r438 and r808?

File: src/fparser/two/Fortran2003.py [hunk 4]
```diff
@@ -5961,6 +6011,7 @@ def tofortran(self, tab='', isfix=None):
             lblock.append(item.tofortran(tab=tab+extra_tab, isfix=isfix))
         if len(self.content) > 1:
             lblock.append(end.tofortran(tab=tab, isfix=isfix))
+        py2_encode_list_items(lblock)
```
Comment (rupertford) [inline src/fparser/two/Fortran2003.py]: Add a test for this as was done for r438 and r808?

File: src/fparser/two/Fortran2003.py [hunk 5]
```diff
@@ -6263,6 +6324,7 @@ def tofortran(self, tab='', isfix=None):
                 extra_tab += '  '
         if len(self.content) > 1:
             line.append(end.tofortran(tab=tab, isfix=isfix))
+        py2_encode_list_items(line)
```
Comment (rupertford) [inline src/fparser/two/Fortran2003.py]: Add a test for this as was done for r438 and r808?

File: src/fparser/two/tests/fortran2003/test_case_construct_r808.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+''' pytest module for the Fortran2003 Case Construct - R808.'''
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2003/test_case_construct_r808.py]: Please add a comment explaining that this only tests part of r808

File: src/fparser/two/tests/fortran2003/test_component_part_r438.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+'''Test Fortran 2003 rule R438: This file tests the support for the components
+of a derived type.
+
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2003/test_component_part_r438.py]: Please list the constraints associated with this rule that are not tested here.

File: src/fparser/two/tests/fortran2003/test_component_part_r438.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    ''' Test that various data component declarations are
+    recognised. R440. '''
+    code = var_type + " :: iflag"
+    reader = FortranStringReader(code, ignore_comments=False)
+    # Ensure reader in in 'free-format' mode
+    reader.set_format(FortranFormat(True, False))
+    obj = Component_Part(reader)
+    assert "iflag" in str(obj)
+
+
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2003/test_component_part_r438.py]: Please add tests that ensures no match when a) data component part and b) proc_component part are invalid.

File: src/fparser/two/tests/test_bases.py [hunk 9]
```diff
@@ -114,3 +114,21 @@ def test_io_ctrl_spec_errors():
     # description
     obj = Io_Control_Spec.match("not_unit=23")
     assert obj is None
+
+
+def test_blockbase_tofortran_non_ascii(f2003_create):
```
Comment (rupertford) [inline src/fparser/two/tests/test_bases.py]: It's unclear what part of blockbase this test exercises. Could you please make it more clear, preferably explicitly calling the blockbase method.

File: src/fparser/two/utils.py [hunk 10]
```diff
@@ -650,6 +669,8 @@ def tofortran(self, tab='', isfix=None):
             mylist.append(item.tofortran(tab=tab+extra_tab, isfix=isfix))
         if len(self.content) > 1:
             mylist.append(end.tofortran(tab=tab, isfix=isfix))
+        # Ensure all strings in list are encoded consistently
+        py2_encode_list_items(mylist)
```
Comment (rupertford) [inline src/fparser/two/utils.py]: Is there a test for this?

File: src/fparser/two/Fortran2003.py (hunk 1 above)
Comment (arporter) [reply src/fparser/two/Fortran2003.py]: Done.

File: src/fparser/two/Fortran2003.py (hunk 2 above)
Comment (arporter) [reply src/fparser/two/Fortran2003.py]: Yes. I've also moved the associated where-construct test out of the test_fortran2003.py file.

File: src/fparser/two/Fortran2003.py (hunk 3 above)
Comment (arporter) [reply src/fparser/two/Fortran2003.py]: Yes, done.

File: src/fparser/two/Fortran2003.py (hunk 4 above)
Comment (arporter) [reply src/fparser/two/Fortran2003.py]: Done.

File: src/fparser/two/Fortran2003.py (hunk 5 above)
Comment (arporter) [reply src/fparser/two/Fortran2003.py]: Done.

File: src/fparser/two/tests/fortran2003/test_case_construct_r808.py (hunk 6 above)
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_case_construct_r808.py]: Done although I've now brought in the related test from test_fortran2003.py.

File: src/fparser/two/tests/fortran2003/test_component_part_r438.py (hunk 7 above)
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_component_part_r438.py]: Done.

File: src/fparser/two/tests/fortran2003/test_component_part_r438.py (hunk 8 above)
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_component_part_r438.py]: Done.

File: src/fparser/two/tests/test_bases.py (hunk 9 above)
Comment (arporter) [reply src/fparser/two/tests/test_bases.py]: In fact, it was just repeating the test from Case_Construct so I've removed it altogether.

File: src/fparser/two/utils.py (hunk 10 above)
Comment (arporter) [reply src/fparser/two/utils.py]: Yes - it was the one that I've just said I've removed. I've put it back and made it call this method explicitly.

**Comment by arporter** [comment]
Ready for second review. I've discovered that `fparser.api.get_reader()` doesn't like non-ASCII chars either but have punted that to a separate Issue.

File: src/fparser/two/tests/test_bases.py [hunk 11]
```diff
@@ -114,3 +114,25 @@ def test_io_ctrl_spec_errors():
...
```
Comment (rupertford) [inline src/fparser/two/tests/test_bases.py]: Make use of `@pytest.mark.usefixtures("f2003_create")` here to be consistent?

Comment (arporter) [reply src/fparser/two/tests/test_bases.py]: Yes. In fact I've done throughout and removed the old-style creation of a parser object from the start of the file.

**Comment by arporter** [comment]
I've created the new Issue as requested (#232) and referenced it in the code and tests. I've also tidied for pylint. Ready for next review.

**Comment by arporter** [comment]
Arrgh. The code I've tidied for pylint isn't covered :-( Will fix.

**Comment by arporter** [comment]
Right. Really is ready for review now.

File: src/fparser/two/tests/fortran2003/test_block_do_construct_r826.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+''' pytest module for the Fortran2003 Do Construct - R826.'''
+
+import pytest
+from fparser.api import get_reader
+from fparser.common.readfortran import FortranStringReader
+from fparser.two.Fortran2003 import Block_Label_Do_Construct, \
+    Block_Nonlabel_Do_Construct
+
+
+@pytest.mark.usefixtures("f2003_create")
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2003/test_block_do_construct_r826.py]: Should this be here?

Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_block_do_construct_r826.py]: Oops. It should but the following two blank lines shouldn't be. There should also have been the same `usefixtures` on the subsequent test (which failed if ran in isolation). I've fixed these problems.

**Pull request merged**